src/theory/quantifiers/trigger.h

```diff
@@ -33,102 +32,180 @@ namespace inst {
+  /** get inst pattern list associated with this trigger */
   Node getInstPattern();
-
-  /** add all available instantiations exhaustively, in any equivalence class
-      if limitInst>0, limitInst is the max # of instantiations to try */
-  int addInstantiations( InstMatch& baseMatch );
+  /* A heuristic value indicating how active this generator is. */
+  int getActiveScore();
+  /** print debug information for the trigger */
+  void debugPrint( const char * c ) {
+    Trace(c) << "TRIGGER( ";
+    for( int i=0; i<(int)d_nodes.size(); i++ ){
+      if( i>0 ){ Trace(c) << ", "; }
+      Trace(c) << d_nodes[i];
+    }
+    Trace(c) << " )";
+  }
+public:
   /** mkTrigger method
-     ie     : quantifier engine;
-     f      : forall something ....
-     nodes  : (multi-)trigger
-     keepAll: don't remove unneeded patterns;
-     trOption : policy for dealing with triggers that already existed
-                (see below)
+  *  qe     : quantifier engine;
+  *  q      : the quantified formula we are making a trigger for
+  *  nodes  : the nodes comprising the (multi-)trigger
+  *  keepAll: don't remove unneeded patterns;
+  *  trOption : policy for dealing with triggers that already existed
+  *             (see below)
+  *  use_n_vars : number of variables that should be bound by the trigger
+  *               typically, the number of quantified variables in f.
   */
   enum{
     TR_MAKE_NEW,    //make new trigger even if it already may exist
     TR_GET_OLD,     //return a previous trigger if it had already been created
     TR_RETURN_NULL  //return null if a duplicate is found
   };
-  //nodes input, trNodes output
-  static bool mkTriggerTerms( Node q, std::vector< Node >& nodes, unsigned n_vars, std::vector< Node >& trNodes );
-  static Trigger* mkTrigger( QuantifiersEngine* qe, Node f, std::vector< Node >& nodes,
+  static Trigger* mkTrigger( QuantifiersEngine* qe, Node q, std::vector< Node >& nodes,
                              bool keepAll = true, int trOption = TR_MAKE_NEW, unsigned use_n_vars = 0 );
-  static Trigger* mkTrigger( QuantifiersEngine* qe, Node f, Node n, bool keepAll = true,
+  /** single trigger version that calls the above function */
+  static Trigger* mkTrigger( QuantifiersEngine* qe, Node q, Node n, bool keepAll = true,
                              int trOption = TR_MAKE_NEW, unsigned use_n_vars = 0 );
+  /** make trigger terms 
+  * This takes a set of eligible trigger terms nodes and returns a subset trNodes such that :
+  *   (1) the terms in trNodes contains at least n_vars of the quantified variables in quantified formula q
+  *   (2) the set trNodes is minimal, i.e. removing one term from trNodes always violates (1).
+  */
+  static bool mkTriggerTerms( Node q, std::vector< Node >& nodes, unsigned n_vars, std::vector< Node >& trNodes );
+  /** collectPatTerms
+  * This collects all terms that are eligible for triggers for quantified formula q in term n.
+  *   tstrt is the selection strategy
```

Comment:
explain what the selection strategies are and how they work - should the selection strategy enum be here?

Reply:
This option is publicly available, so I think it needs to be in options/quantifiers_modes.h